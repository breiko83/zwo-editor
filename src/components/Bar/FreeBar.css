.segment {
  position: relative;
}
.freeRide {    
  border-radius: 0px;
  border: 1px solid white;
  background-image: url('../../assets/freeride.svg');
}