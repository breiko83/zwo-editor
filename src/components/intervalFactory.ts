import { v4 as uuidv4 } from 'uuid'
import { PaceType } from './Editor/PaceSelector';
import { FreeInterval, Interval, RampInterval, RepetitionInterval, SteadyInterval } from "./Interval";

const defaultDuration = 300;
const defaultPower = 1.0;
const defaultPace = PaceType.oneMile;
const defaultCadence = 0;

export default {
  steady(interval: Partial<SteadyInterval>): SteadyInterval {
    return {
      type: 'steady',
      id: uuidv4(),
      duration: defaultDuration,
      power: defaultPower,
      cadence: defaultCadence,
      pace: defaultPace,
      ...interval,
    };
  },

  ramp(interval: Partial<RampInterval>): RampInterval {
    return {
      type: 'ramp',
      id: uuidv4(),
      duration: defaultDuration,
      startPower: defaultPower / 2,
      endPower: defaultPower,
      cadence: defaultCadence,
      pace: defaultPace,
      ...interval,
    };
  },

  free(interval: Partial<FreeInterval>): FreeInterval {
    return {
      type: 'free',
      id: uuidv4(),
      duration: defaultDuration,
      cadence: defaultCadence,
      ...interval
    };
  },

  repetition(interval: Partial<RepetitionInterval>): RepetitionInterval {
    return {
      type: 'repetition',
      id: uuidv4(),
      cadence: defaultCadence,
      restingCadence: defaultCadence,
      repeat: 3,
      onDuration: 30,
      offDuration: 120,
      onPower: defaultPower,
      offPower: defaultPower / 2,
      pace: defaultPace,
      ...interval,
    };
  },

  // Creates a copy of the interval, with a new ID
  clone(interval: Interval): Interval {
    return { ...interval, id: uuidv4() };
  },
};
