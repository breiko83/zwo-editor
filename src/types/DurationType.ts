export type DurationType = "time" | "distance";