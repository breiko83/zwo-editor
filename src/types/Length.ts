export class Duration {
  constructor(public readonly seconds: number) {}
}

export class Distance {
  constructor(public readonly meters: number) {}
}

export type Length = Distance | Duration;
