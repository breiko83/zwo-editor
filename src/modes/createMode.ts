import { LengthType } from "../types/LengthType";
import { RunningTimes } from "../types/RunningTimes";
import { SportType } from "../types/SportType";
import BikeMode from "./BikeMode";
import RunMode from "./RunMode";
import { WorkoutMode } from "./WorkoutMode";

interface CreateModeParams {
  sportType: SportType;
  ftp: number;
  weight: number;
  runningTimes: RunningTimes;
  lengthType: LengthType;
}

// Creates appropriate type of mode for current sport type
export default function createMode({sportType, ftp, weight, runningTimes, lengthType}: CreateModeParams): WorkoutMode {
  switch (sportType) {
    case "bike": return new BikeMode(ftp, weight);
    case "run": return new RunMode(runningTimes, lengthType);
  }
}
