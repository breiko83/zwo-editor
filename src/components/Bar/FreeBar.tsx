import React, { useState } from 'react'
import './FreeBar.css'
import { Resizable } from 're-resizable'
import 'moment-duration-format'
import Label from '../Label/Label'
import { FreeInterval } from '../../types/Interval'
import { durationMultiplier } from './multipliers'
import { WorkoutMode } from '../../modes/WorkoutMode'

interface FreeBarProps {
  interval: FreeInterval;
  mode: WorkoutMode;
  onChange: (interval: FreeInterval) => void;
  onClick: (id: string) => void;
  selected: boolean;
}

const FreeBar = ({interval, mode, ...props}: FreeBarProps) => {
  const [width, setWidth] = useState(mode.lengthToWidth(interval.length))

  const [showLabel, setShowLabel] = useState(false)

  const handleCadenceChange = (cadence: number) => {
    props.onChange({ ...interval, cadence })
  }

  // standard height
  const height = 100

  const handleResizeStop = (dWidth: number) => {
    setWidth(width + dWidth)
    notifyChange(dWidth)
  }

  const notifyChange = (dWidth: number) => {
    props.onChange({ ...interval, length: mode.widthToLength(width + dWidth) })
  }

  return (
    <div className='segment'
      onMouseEnter={() => setShowLabel(true)}
      onMouseLeave={() => setShowLabel(false)}
      style={props.selected ? {zIndex:1}: {}}
      onClick={() => props.onClick(interval.id)}
    >
      {(props.selected || showLabel) &&
        <Label
          interval={interval}
          mode={mode}
          onCadenceChange={(cadence: number)=> handleCadenceChange(cadence)}
        />
      }
      <Resizable
        className='freeRide'
        size={{
          width: mode.lengthToWidth(interval.length),
          height: height,
        }}
        minWidth={durationMultiplier}
        minHeight={height}
        maxHeight={height}
        enable={{ right: true }}
        grid={[1, 1]}
        onResizeStop={(e, direction, ref, d) => handleResizeStop(d.width)}
        onResize={(e, direction, ref, d) => notifyChange(d.width)}        
      >
      </Resizable>
    </div>
  )

}

export default FreeBar
