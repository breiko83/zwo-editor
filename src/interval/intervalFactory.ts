import { v4 as uuidv4 } from 'uuid'
import { PaceType } from '../types/PaceType';
import { FreeInterval, Interval, RampInterval, RepetitionInterval, SteadyInterval } from "../types/Interval";
import { Distance, Duration, Length } from '../types/Length';
import { WorkoutMode } from '../modes/WorkoutMode';

const defaultDuration = new Duration(300);
const defaultDistance = new Distance(1000);
const defaultIntensity = 1.0;
const defaultPace = PaceType.oneMile;
const defaultCadence = 0;

const defaultLength = (mode: WorkoutMode): Length =>
  mode.lengthType === "time" ? defaultDuration : defaultDistance;

export default {
  steady(interval: Partial<SteadyInterval>, mode: WorkoutMode): SteadyInterval {
    return {
      type: 'steady',
      id: uuidv4(),
      length: defaultLength(mode),
      intensity: defaultIntensity,
      cadence: defaultCadence,
      pace: defaultPace,
      ...interval,
    };
  },

  ramp(interval: Partial<RampInterval>, mode: WorkoutMode): RampInterval {
    return {
      type: 'ramp',
      id: uuidv4(),
      length: defaultLength(mode),
      startIntensity: defaultIntensity / 2,
      endIntensity: defaultIntensity,
      cadence: defaultCadence,
      pace: defaultPace,
      ...interval,
    };
  },

  free(interval: Partial<FreeInterval>, mode: WorkoutMode): FreeInterval {
    return {
      type: 'free',
      id: uuidv4(),
      length: defaultLength(mode),
      cadence: defaultCadence,
      ...interval
    };
  },

  repetition(interval: Partial<RepetitionInterval>, mode: WorkoutMode): RepetitionInterval {
    return {
      type: 'repetition',
      id: uuidv4(),
      onCadence: defaultCadence,
      offCadence: defaultCadence,
      repeat: 3,
      onLength: mode.lengthType === "time" ? new Duration(30) : new Distance(200),
      offLength: mode.lengthType === "time" ? new Duration(120) : new Distance(200),
      onIntensity: defaultIntensity,
      offIntensity: defaultIntensity / 2,
      pace: defaultPace,
      ...interval,
    };
  },

  // Creates a copy of the interval, with a new ID
  clone(interval: Interval): Interval {
    return { ...interval, id: uuidv4() };
  },
};
