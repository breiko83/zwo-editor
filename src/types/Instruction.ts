import { v4 as uuidv4 } from "uuid";
import { WorkoutMode } from "../modes/WorkoutMode";
import { Distance, Duration, Length } from "./Length";

export interface Instruction {
  id: string,
  text: string,
  offset: Length,
}

export function createInstruction(fields: Partial<Instruction>, mode: WorkoutMode): Instruction {
  return {
    id: uuidv4(),
    text: "",
    offset: mode.lengthType === "time" ? new Duration(0) : new Distance(0),
    ...fields,
  };
}
