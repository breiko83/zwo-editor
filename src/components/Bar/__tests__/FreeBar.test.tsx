import React from 'react';
import FreeBar from '../FreeBar';
import renderer from 'react-test-renderer';
import '@testing-library/jest-dom/extend-expect'
import intervalFactory from '../../../interval/intervalFactory';
import createMode from '../../../modes/createMode';
import { Duration } from '../../../types/Length';

const MockReact = React;

jest.mock('../../Label/Label', () => (props: any) =>
  MockReact.createElement("Label", props));

jest.mock('re-resizable', () => ({
  Resizable: (props: any) => MockReact.createElement("Resizable", props),
}));

jest.mock('uuid', () => ({
  v4: () => `mock-id`,
}));

describe('<FreeBar>', () => {
  it('renders', () => {
    const mode = createMode({sportType: "bike", ftp: 200, weight: 75, runningTimes: [], lengthType: "time"});
    const interval = intervalFactory.free({
      length: new Duration(50),
    }, mode);
  
    const component = renderer.create(
      <FreeBar
        interval={interval}
        mode={mode}
        selected={false}
        onChange={() => { }}
        onClick={() => { }}
      />
    );
  
    expect(component).toMatchSnapshot();
  });

  it('renders with label when selected', () => {
    const mode = createMode({sportType: "bike", ftp: 200, weight: 75, runningTimes: [], lengthType: "time"});
    const interval = intervalFactory.free({
      length: new Duration(50),
    }, mode);
  
    const component = renderer.create(
      <FreeBar
        interval={interval}
        mode={mode}
        selected={true}
        onChange={() => { }}
        onClick={() => { }}
      />
    );
  
    expect(component).toMatchSnapshot();
  });
});
