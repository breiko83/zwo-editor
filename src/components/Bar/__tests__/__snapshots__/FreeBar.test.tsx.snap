// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FreeBar> renders 1`] = `
<div
  className="segment"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  style={Object {}}
>
  <Resizable
    className="freeRide"
    enable={
      Object {
        "right": true,
      }
    }
    grid={
      Array [
        1,
        1,
      ]
    }
    maxHeight={100}
    minHeight={100}
    minWidth={3}
    onResize={[Function]}
    onResizeStop={[Function]}
    size={
      Object {
        "height": 100,
        "width": 16.666666666666668,
      }
    }
  />
</div>
`;

exports[`<FreeBar> renders with label when selected 1`] = `
<div
  className="segment"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  style={
    Object {
      "zIndex": 1,
    }
  }
>
  <Label
    interval={
      Object {
        "cadence": 0,
        "id": "mock-id",
        "length": Duration {
          "seconds": 50,
        },
        "type": "free",
      }
    }
    mode={
      BikeMode {
        "ftp": 200,
        "lengthType": "time",
        "sportType": "bike",
        "weight": 75,
      }
    }
    onCadenceChange={[Function]}
  />
  <Resizable
    className="freeRide"
    enable={
      Object {
        "right": true,
      }
    }
    grid={
      Array [
        1,
        1,
      ]
    }
    maxHeight={100}
    minHeight={100}
    minWidth={3}
    onResize={[Function]}
    onResizeStop={[Function]}
    size={
      Object {
        "height": 100,
        "width": 16.666666666666668,
      }
    }
  />
</div>
`;
